environmental/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+//! }
+//! ```
+
+#[doc(hidden)]
+pub use std::cell::RefCell;
+use std::thread::LocalKey;
+
+#[doc(hidden)]
+pub fn using<'a, T: 'a, R, S, F: FnOnce() -> R>(
+	global: &'static LocalKey<RefCell<*mut S>>,
+	protected: &'a mut T,
+	f: F
+) -> R {
+	// store the `protected` reference as a pointer so we can provide it to logic running within
+	// `f`.
+	// while we record this pointer (while it's non-zero) we guarantee:
+	// - it will only be used once at any time (no reentrancy);
+	// - that no other thread will use it; and
+	// - that we do not use the original mutating reference while the pointer.
+	// exists.
+	let original = global.with(|r| {
+		let mut b = r.borrow_mut();
+		let o = *b;
+		*b = protected as *mut T as *mut S;
+		o
+	});
+	let r = f();
+	global.with(|r| *r.borrow_mut() = original);
+	r
+}
+
+#[doc(hidden)]
+pub fn with<'r, R, S, T: 'r, F: FnOnce(&'r mut T) -> R>(
+	global: &'static LocalKey<RefCell<*mut S>>,
+	mutator: F,
+) -> Option<R> {
+	global.with(|r| {
+		let br = r.borrow_mut();
+		if *br != 0 as *mut S {
+			// safe because it's only non-zero when it's being called from using, which
+			// is holding on to the underlying reference (and not using it itself) safely.
+			unsafe {
+				Some(mutator(&mut *(*br as *mut S as *mut T)))
+			}
+		} else {
+			None
+		}
+	})
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! decl {
+	($name:ident : $t:ty) => {
+		thread_local! {
+			static $name: $crate::RefCell<*mut $t> = $crate::RefCell::new(0 as *mut $t);
+		}
+	}
+}
+
+/// Declare a new global reference module whose underlying value does not contain references.
+///
+/// Will create a module of a given name that contains two functions:
+/// * `pub fn using<'a: 'b, 'b, R, F: FnOnce() -> R, T: 'a>(protected: &'b mut T, f: F) -> R`
+///   This executes `f`, returning its value. During the call, the module's reference is set to
+///   be equal to `protected`.
+/// * `pub fn with<R, F: for<'r, 't: 'r> FnOnce(&'r mut $t<'t>) -> R>(f: F) -> Option<R>`
+///   This executes `f`, returning `Some` of its value if called from code that is being executed
+///   as part of a `using` call. If not, it returns `None`. `f` is provided with one argument: the
+///   same reference as provided to the most recent `using` call.
+///
+/// # Examples
+///
+/// ```
+/// #[macro_use] extern crate environmental;
+/// declare_simple!(counter: u32);
+/// fn main() {
+///   let mut counter_value = 41u32;
+///   counter::using(&mut counter_value, || {
+///     let odd = counter::with(|value|
+///       if *value % 2 == 1 {
+///         *value += 1; true
+///       } else {
+///         *value -= 3; false
+///       }).unwrap();	// safe because we're inside a counter::using
+///     println!("counter was {}", match odd { true => "odd", _ => "even" });
+///   });
+///
+///   println!("The answer is {:?}", counter_value); // 42
+/// }
+/// ```
+#[macro_export]
+macro_rules! declare_simple {
+	($name:ident : $t:tt) => {
```

Comment:
should be `$t:ty`, otherwise you can't use types whose names are multiple tokens like `Vec<u8>`